server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java

```diff
@@ -1515,6 +1509,34 @@ public SearchPhase newSearchPhase(
         }
     }
 
+    /**
+     * TransportSearchAction cannot access async-search code, so can't check whether this the Task
+     * is an instance of AsyncSearchTask, so this roundabout method is used
+     * @param searchTask SearchTask to analyze
+     * @return true if this is an async search task; false if a synchronous search task
+     */
+    private boolean isAsyncSearchTask(SearchTask searchTask) {
+        assert assertAsyncSearchTaskListener(searchTask) : "AsyncSearchTask SearchProgressListener is not one of the expected types";
+        // AsyncSearchTask will not return SearchProgressListener.NOOP, since it uses its own progress listener
+        // which delegates to CCSSingleCoordinatorSearchProgressListener when minimizing roundtrips.
+        // Only synchronous SearchTask uses SearchProgressListener.NOOP or CCSSingleCoordinatorSearchProgressListener directly
+        return searchTask.getProgressListener() != SearchProgressListener.NOOP
+            && searchTask.getProgressListener() instanceof CCSSingleCoordinatorSearchProgressListener == false;
```

Comment:
should we expose a new method that provides this information and allows us to remove the instanceof check?

Reply:
This is pre-existing code so I am certainly not against doing that but I don't know how and it doesn't seem to be part of this project. We could add a separate pull for this maybe?

Reply:
It was introduced by @quux00 to figure out whether or not we have async search, I tried to see how to eliminate it but failed to find a better way so far. What would be your suggestion - to add a new method to which class - `SearchProgressListener`?

Reply:
I think Luca means to add a method to SearchTask, since AsyncSearchTask extends it. That would be cleaner I think. Add `public boolean isAsyncSearch()`, where SearchTask returns false and AsyncSearchTask overrides it to return true.

Reply:
Ah that's easy but that means introducing knowledge of async into `SearchTask` which only existed xpack class AsyncSearchTask before. Is that OK?

Reply:
It just adds that one boolean method, so I think that's fine. IMO async search should really be part of server anyway, so there's no reason to pretend server shouldn't know about async_search. It's just historical reasons that we have the module barriers for this case.

Reply:
OK should I add this to this pull or in followup?

Reply:
> OK should I add this to this pull or in followup?

what works best for you.

Reply:
Let's do a followup - this one is pretty big as it is IMHO.